src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
-"trimmed".  There is no way to identify unnecessary nodes in an immutable
-SHAMap that could be removed.  Once a node has been brought into the
-in-memory SHAMap, that node stays in memory for the life of the SHAMap.
+"trimmed".  There is no way to identify unnecessary nodes in an immutable SHAMap
+that could be removed.  Once a node has been brought into the in-memory SHAMap,
+that node stays in memory for the life of the SHAMap.
 
-Most SHAMaps are immutable, in the sense that they don't modify or remove
-their contained nodes.
+Most SHAMaps are immutable, in the sense that they don't modify or remove their
+contained nodes.
 
 An example where a mutable SHAMap is required is when we want to apply
-transactions to the last closed ledger.  To do so we'd make a mutable
-snapshot of the state tree and then start applying transactions to it.
-Because the snapshot is mutable, changes to nodes in the snapshot will not
-affect nodes in other SHAMAps.
+transactions to the last closed ledger.  To do so we'd make a mutable snapshot
+of the state trie and then start applying transactions to it. Because the
+snapshot is mutable, changes to nodes in the snapshot will not affect nodes in
+other SHAMaps.
 
-An example using a immutable ledger would be when there's an open ledger
-and some piece of code wishes to query the state of the ledger.  In this
-case we don't wish to change the state of the SHAMap, so we'd use an
-immutable snapshot.
+An example using a immutable ledger would be when there's an open ledger and
+some piece of code wishes to query the state of the ledger.  In this case we
+don't wish to change the state of the SHAMap, so we'd use an immutable snapshot.
 
+## Sequence numbers ##
 
-## SHAMap Creation ##
+Both SHAMap's and their nodes carry a sequence number.  This is simply an
+unsigned number that indicates ownership or membership, or a non-membership.
 
-A SHAMap is usually not created from vacuum.  Once an initial SHAMap is
-constructed, later SHAMaps are usually created by calling
-snapShot(bool isMutable) on the original SHAMap().  The returned SHAMap
-has the expected characteristics (mutable or immutable) based on the passed
-in flag.
+SHAMap's sequence numbers normally start out as 1.  However when a snap-shot of
+a SHAMap is made, the copy's sequence number is 1 greater than the original.
 
-It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
-snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might
-be modified must be copied before they are placed in the mutable map.
+The nodes of a SHAMap have their own copy of a sequence number.  If the SHAMap
+is not immutable, meaning it can change, then all of its nodes must have the
+same sequence number as the SHAMap itself.  This enforces an invariant that none
+of the nodes are shared with other SHAMaps.
 
+When a SHAMap gets a copy of a node from elsewhere, and must have a private
```

Comment:
What does it mean for a SHAMap to get a copy of a node from elsewhere? When would this happen and where is the elsewhere? Why would the SHAMap need a private copy? I think it would be useful to add an example here.